### ISSUE TITLE
Be able to regexp filter tree

### ISSUE THREAD

**Issue by nacesprin** [Gruntfuggly/todo-tree#659]
![image](https://user-images.githubusercontent.com/295375/182093179-931081c4-705b-4b1e-9ca0-c9ad3bc333b9.png)

At now we only filter by one word. It would be handly to filter for more than one word. For example: TODO and CONTINUE options, or whatever I want.

**Comment by Gruntfuggly** [comment]
This might be a reasonably simple enhancement, so I'll try and add it.

**Comment by Gruntfuggly** [comment]
Not sure when I implemented this, but it already works. E.g. `(TODO|CONTINUE)` should work as you expect.

**Comment by nacesprin** [comment]
TL;DR; the filter is filtering by words in the TODO comment line, but not by TODO keyword

The filtering option is not working as expected. Steps to reproduce:

1) This is my TODO tab:
![image](https://user-images.githubusercontent.com/295375/217192356-a53537e9-72b6-45bb-accb-347df5174840.png)

2) Filtering by TODO keyword.
Expected: view TODO keyword results.
Shown: nothing, because there is not any comment which contains TODO **word**
![image](https://user-images.githubusercontent.com/295375/217192463-398c53a8-5188-40c0-b022-dffd72f51a71.png)

![image](https://user-images.githubusercontent.com/295375/217192840-f3fd5984-9859-44a7-8fd9-cab49f3c636e.png)

3) Filtering by _**something**_ word
Expected: nothing, because there is no _**something**_ keyword
Shown: results by containing _**something**_ word. It should not be thrown because _**something**_ is a word of a comment, not a KEYWORD
![image](https://user-images.githubusercontent.com/295375/217194548-816c6f8b-e0d2-4cd9-bf57-5c810732abcf.png)

![image](https://user-images.githubusercontent.com/295375/217193352-9c4986a5-0875-46bb-9862-7c20f9188680.png)

**Comment by Gruntfuggly** [comment]
Ah I see - yes, at the moment it only includes the leaf nodes, not the folder nodes. So what would be the desired behaviour? If it matched a keyword, would you expect all it's children to be shown or just the node itself?

**Comment by nacesprin** [comment]
Thanks Nigel. I suppose a checkbox inside the filter dialog to choose between "folder nodes" / "leaf nodes" options would be handy. Then depending of the user selection, will filter it.

**Comment by nacesprin** [comment]
Or another 3rd option like "both" to filter in both ways.

**Comment by frypf** [comment]
> TL;DR; the filter is filtering by words in the TODO comment line, but not by TODO keyword

I only witness this when grouping by tag. I don't tend to use that option normally - even less now since I've realised I could filter by tag - so it seemed to be working fine as far as I could tell until I played around a bit with different options.

On a related note, is there any way to pass a regex as a string arg directly to the `todo-tree.filter` command? Being able to assign a keybinding to quickly bring up a list of solely TODO items / project marks / HACKs or whatever without having to type out the different regexes each time would be super-useful.

**Comment by frypf** [comment]
Had a quick test on a local build, and on the face of it making a small change [here](https://github.com/Gruntfuggly/todo-tree/blob/84ccc168e84752c9b7064329f2abe37335764874/src/extension.js#L1273) seems to do the trick wrt passing an arg directly to the filter:
```diff
context.subscriptions.push( vscode.commands.registerCommand( 'todo-tree.filter', function( argTerm )
{
-    vscode.window.showInputBox( { prompt: "Filter tree" } ).then(
+    Promise.resolve(argTerm ?? vscode.window.showInputBox( { prompt: "Filter tree" } )).then(
        function( term )
```

I'm not sure how that might play out with the proposed addition of the checkbox idea though.

**Comment by Gruntfuggly** [comment]
I don't think adding a checkbox is possible - the API doesn't provide anything for that, so it would probably need to be a config option.

What I meant before was, what should the tree show? Say you have the following:
```
  TODO
     item 1
     item 2
```
Currently filtering on "2" would show the tree like
```
  TODO
    item 2
```

If it includes the tag in the filter (e.g. set the filter to "todo") what would you expect to be shown? It could show all the items under the TAG (as if there was no filter) but then how far down do you go? 

Or it could just show the 'TODO' tag on its own, but that seems a bit pointless?

Perhaps it should be a separate filter command altogether - "Filter Tree Tags"  - which would just apply the filter to the folder nodes?

**Comment by Gruntfuggly** [comment]
I'll see if I can add the argument - seems like a simple thing to add.

**Comment by frypf** [comment]
> I'll see if I can add the argument - seems like a simple thing to add.

That'd be awesome, thanks very much. I looked at it again this afternoon and was wondering if it might also be possible to add an extra setting `"todo-tree.tree.showActiveFilters"`, (similar to `"todo-tree.tree.showCurrentScanMode"`) to optionally remove the `filterStatusNode` from the list. I like the idea of being able to navigate the list with cursor keys as discussed in #730 (I already have something similar working via a macro with [vscode-commands](https://github.com/usernamehw/vscode-commands)), and I feel this would work really well in consort with that proposed change.
In terms of making sure the list still shows when a filter is active, the filter button already changes to indicate this, but I've also tinkered with making sure that the node is always shown if an active filter fails to match anything. I'm happy enough to attempt a pull request or just post some diffs if you feel this might work - the changes are pretty minimal.

> I don't think adding a checkbox is possible

Having looked again at the API doc today, I was coming to the same conclusion. I'd be fine with separate commands, not sure how it may suit other peoples' use-cases? I was wondering is there any way of fuzzy-searching (live-filtering as you type) within a list view, or is that only available to the quick picker?

**Comment by frypf** [comment]
I had a look again at adding `filterTags` as a separate command - it struck me it might be better to do this one as a straight `tagList.includes(node.tag)` rather than applying a regex comparison. Please feel free to look at [my fork](https://github.com/frypf/todo-tree) if it's of any interest.
